fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/expressions/RelationalExpression.java

```diff
@@ -729,6 +749,11 @@ default Optional<MatchInfo> adjustMatch(@Nonnull final PartialMatch partialMatch
         return Optional.empty();
     }
 
+    default boolean isCompatiblyAndCompletelyBound(@Nonnull final AliasMap bindingAliasMap, @Nonnull final List<? extends Quantifier> candidateQuantifiers) {
+        return !hasUnboundQuantifiers(bindingAliasMap) &&
+                !hasIncompatibleBoundQuantifiers(bindingAliasMap, candidateQuantifiers);
```

Comment:
Hmm, so `SelectExpression` can and do match existential quantifiers against for-each quantifiers. That's why this is not used in `SelectExpression` and therefore tolerated. Only expressions like group-by,  etc. need a for-each to a for-each. Since we don't really allow an existential in any other expression than a select the point is probably moot and you are right, however, let's keep it like this since this is more future-proof.